cf/dimensioncoordinate.py

```diff
@@ -908,6 +1011,164 @@ def roll(self, axis, shift, inplace=False, i=False):
+        >>> c.set_bounds(b)
+        >>> c.has_bounds()
+        True
+        >>> c.get_bounds()
+        <Bounds: (5, 2) >
+        >>> b = c.del_bounds()
+        >>> b
+        <Bounds: (5, 2) >
+        >>> c.has_bounds()
+        False
+        >>> print(c.get_bounds(None))
+        None
+        >>> print(c.del_bounds(None))
+        None
+
+        """
+        self._del_component("cell_characteristics", default=None)
+        super().set_bounds(bounds, copy=copy)
+
+    def set_cell_characteristics(self, cellsize, spacing):
+        """Set cell characteristics.
+
+        A cell characterisitic is assumed to valid for each cell. Cell
+        characteristics are not inferred from the coordinate or bounds
+        data, but may be set with this method. Cell characteristics
+        are automatically removed whenever the new data or bounds are
+        set with `set_data` or `set_bounds` respectively.
+
+        .. versionadded:: 3.15.4
+
+        .. seealso:: `del_cell_characteristics`,
+                     `get_cell_characteristics`,
+                     `has_cell_characteristics`
+
+        :Parameters:
+
+            cellsize:
+                The cell size (i.e. the absolute difference between
+                the cell bounds) characteristic. May be a `Query`,
+                `TimeDuration`, scalar `Data`, scalar data_like
+                object, or `None`. A value of `None` means no
+                characteristc has been set.
+
+            spacing:
+                The cell spacing (i.e. the absolute difference between
+                two neighbouring coordinate values) characteristic.
+                May be a `Query`, `TimeDuration`, scalar `Data`,
+                scalar data_like object, or `None`. A value of `None`
+                means no characteristc has been set.
+
+        :Returns:
+
+            `None`
+
+        **Examples**
+
+        >>> d.set_cell_characteristics(cellsize=cf.D(5), spacing=cf.D(1))
+
+        >>> d.set_cell_characteristics(cf.Data(10, 'degree_E'), None)
+
+        >>> d.set_cell_characteristics(cf.wi(100, 200), 150)
+
+        """
+        chars = {}
+        if cellsize is not None:
+            chars["cellsize"] = cellsize
+
+        if spacing is not None:
+            chars["spacing"] = spacing
+
+        if not chars:
+            self.del_cell_characteristics(None)
+            return
+
+        self._set_component("cell_characteristics", chars, copy=False)
+
+    def set_data(self, data, copy=True, inplace=True):
+        """Set the data.
+
+        The units, calendar and fill value of the incoming `Data`
+        instance are removed prior to insertion.
+
+        .. versionadded:: 3.15.4
+
+        .. seealso:: `data`, `del_data`, `get_data`, `has_data`
+
+        :Parameters:
+
+            data: `Data`
+                The data to be inserted.
+
+                {{data_like}}
+
+            copy: `bool`, optional
+                If False then do not copy the data prior to
```

Comment:
Similar to a previous suggestion:

```suggestion
                If True then copy the data prior to
```